UefiPayloadPkg/Library/PciHostBridgeLib/PciHostBridgeSupport.c

```diff
@@ -596,3 +597,67 @@ RetrieveRootBridgeInfoFromHob (
 
   return PciRootBridges;
 }
+
+static PCI_SEGMENT_INFO  *PciSegments = 0;
+
+/**
+  Find segment info from all root bridges
+
+  @param[in]  PciRootBridgeInfo    Pointer of Universal Payload PCI Root Bridge Info Hob
+  @param[out] NumberOfRootBridges  Number of root bridges detected
+
+  @retval     Pointer to the allocated Segment info structure array.
+
+**/
+PCI_SEGMENT_INFO *
+RetrieveSegmentInfoFromHob (
+  IN  UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES  *PciRootBridgeInfo,
+  OUT UINTN                               *NumberOfRootBridges
+  )
+{
+  UINTN  Size;
+  UINT8  Index;
+
+  ASSERT (PciRootBridgeInfo != NULL);
+  ASSERT (NumberOfRootBridges != NULL);
+  if (PciRootBridgeInfo == NULL) {
+    return NULL;
+  }
+
+  if (PciRootBridgeInfo->Count == 0) {
+    PcdSetBoolS (PcdPciDisableBusEnumeration, FALSE);
+    return NULL;
+  }
+
+  *NumberOfRootBridges = PciRootBridgeInfo->Count;
+  if (PciSegments != NULL) {
+    return PciSegments;
+  }
+
+  Size        = PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO);
+  PciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
+  ASSERT (PciSegments != NULL);
+  ZeroMem (PciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
+
+  //
+  // Create all root bridges with PciRootBridgeInfoHob
+  //
+  for (Index = 0; Index < PciRootBridgeInfo->Count; Index++) {
+    PciSegments[Index].BaseAddress    = PciRootBridgeInfo->RootBridge[Index].Ecam.Base;
+    PciSegments[Index].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment);
+    PciSegments[Index].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Base;
+    PciSegments[Index].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Limit;
+  }
+
+  //
+  // Now, this library only supports RootBridge that ResourceAssigned is True
+  //
+  if (PciRootBridgeInfo->ResourceAssigned) {

```

Comment:
Do we really need to set PcdPciDisableBusEnumeration here? 
GetPciSegmentInfo() may be called every time PciSegmentRead/Write called, so we may not like to toggle PCD values in such frequently calling function.

Reply:
I agree. Removing PCD setting here.

Reply:
resolved